### ISSUE TITLE
General convolution/deconvolution solution

### ISSUE THREAD

**Issue by Wangyida** [tiny-dnn/tiny-dnn#303]
I want to modify the conv and deconv pipeline based on this concept [conv_arithmetic](https://github.com/vdumoulin/conv_arithmetic) rather than my previous one. My previous deconvolution treat forward pass operation in a single input unit as **input(1*1) _conv_ filter(n*n) = output(n*n)** while the transposed convolution treat it in another way similar to convolution as **input(n*n) _conv_ filter(n*n) = output(1*1)**. The result is the same when the filters are transposed in all elements and the output result is only determined by padding and striding strategy.
@bhack do you think this is better? Seems a general convolution related topic.

**Comment by bhack** [comment]
Yes, see expecially 4.2 section of https://arxiv.org/abs/1603.07285

**Comment by nyanp** [comment]
@Wangyida 
It's very good idea to unify conv/deconv operations. There are too many copy&paste in current implementations and it makes tiny-dnn difficult to maintain. Do you have actual plan to refactor them?

**Comment by wangyida** [comment]
@bhack I am currently have a simple test on that paper using transposed conv to replace deconv just as described in [A guide to convolution arithmetic for deep
learning](https://github.com/vdumoulin/conv_arithmetic) and the [paper](https://arxiv.org/abs/1603.07285) you shared with me, there are no problem in regular deconv with stride equals to 1. Do you think it's OK for transposed conv with other strides described there?
@nyanp @bhack I think it's rather simple to treat deconv and conv in this way because here I should just treat the input size and input data for both conv and deconv, so I can just reuse the codes for computational kernels of conv on deconv in a transposed way. Do you think this way is okay?

**Comment by bhack** [comment]
@wangyida You can see a reference implementation athttps://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/conv.py

**Comment by bhack** [comment]
What is the status of this?
